### PR TITLE
Add local_response_norm in nn.functional and nn.layer

### PR DESCRIPTION
**Pull request by huangjun12**
### PR types
New features

### PR changes
APIs

### Describe
原API:  
paddle.fluid.layers.lrn(input, n=5, k=1.0, alpha=0.0001, beta=0.75, name=None, data_format='NCHW')
存在如下问题:
(1) 精度与torch实现有较大diff，最大相对误差约为1e-4
(2) 仅支持4-D输入，不支持3D及5D输入
(3) api名称、参数名称、参数顺序与torch的接口不同

在functional和layer下新增如下两个API，修复了以上问题，实现和接口形式对齐torch:
(1) paddle.nn.functional.local_response_norm(x, size, alpha=1e-4, beta=0.75, k=1., data_format="NCHW", name=None)
(2) paddle.nn.LocalResponseNorm(size, alpha=0.0001, beta=0.75, k=1.0, data_format="NCHW", name=None)

### CODE REVIEW

File: python/paddle/nn/functional/norm.py [hunk 1]
```diff
@@ -403,3 +402,118 @@ def instance_norm(x,
...
+
+def local_response_norm(x,
+                        size,
+                        alpha=1e-4,
+                        beta=0.75,
+                        k=1.,
+                        data_format="NCHW",
+                        name=None):
+    """
+        Local Response Normalization performs a type of "lateral inhibition" by normalizing over local input regions.
```
Comment (shippingwang) [inline python/paddle/nn/functional/norm.py:415]: refer to its own class doc

File: python/paddle/fluid/tests/unittests/test_lrn_op.py [hunk 2]
```diff
@@ -154,5 +155,207 @@ def test_errors(self):
...
+            self.places.append(fluid.CUDAPlace(0))
+
+    def check_static_3d_input(self, place):
+        with fluid.program_guard(fluid.Program(), fluid.Program()):
+            in_np1 = np.random.random([3, 40, 40]).astype("float32")
+            in_np2 = np.transpose(in_np1, (0, 2, 1))
+
+            input1 = fluid.data(
+                name="input1", shape=[3, 40, 40], dtype="float32")
+            input2 = fluid.data(
```
Comment (shippingwang) [inline python/paddle/fluid/tests/unittests/test_lrn_op.py:172]: input1/2 shape should be [3, 3, 40] and [3, 40, 3]?

File: python/paddle/fluid/tests/unittests/test_lrn_op.py [hunk 3]
```diff
@@ -154,5 +155,207 @@ def test_errors(self):
...
+            fetches1_tran = np.transpose(fetches[1], (0, 4, 1, 2, 3))
+            self.assertTrue(np.allclose(fetches[0], fetches1_tran))
+
+    def test_static(self):
+        for place in self.places:
+            self.check_static_3d_input(place=place)
+            self.check_static_4d_input(place=place)
+            self.check_static_5d_input(place=place)
+
+    def check_dygraph_3d_input(self, place):
```
Comment (shippingwang) [inline python/paddle/fluid/tests/unittests/test_lrn_op.py:240]: so it only check the data_format?

File: python/paddle/fluid/tests/unittests/test_lrn_op.py [hunk 4]
```diff
@@ -154,5 +155,207 @@ def test_errors(self):
...
+
+class TestLocalResponseNormCAPI():
+    def setUp(self):
+        np.random.seed(123)
+        self.places = [fluid.CPUPlace()]
+        if core.is_compiled_with_cuda():
+            self.places.append(fluid.CUDAPlace(0))
+
+    def test_dygraph(self):
+        for place in self.places:
```
Comment (shippingwang) [inline python/paddle/fluid/tests/unittests/test_lrn_op.py:335]: why don't use the new api to test dygraph?

File: python/paddle/nn/functional/norm.py [hunk 5]
```diff
@@ -403,3 +402,118 @@ def instance_norm(x,
...
+    sizes = x.shape
+    dim = len(sizes)
+    if dim < 3:
+        raise ValueError(
+            'Expected 3D or higher dimensionality input, but got {} dimensions'.
+            format(dim))
+
+    channel_last = True if data_format[-1] == "C" else False
+
+    div = fluid.layers.unsqueeze(paddle.multiply(x, x), axes=[1])
```
Comment (willthefrog) [inline python/paddle/nn/functional/norm.py]: use functional instead of layers?

File: python/paddle/nn/functional/norm.py [hunk 6]
```diff
@@ -403,3 +402,118 @@ def instance_norm(x,
...
+    if channel_last:
+        if dim == 3:
```
Comment (willthefrog) [inline python/paddle/nn/functional/norm.py]: this part is a bit too repetitive, could use some consolidation.

**Review by willthefrog — CHANGES_REQUESTED** [review]
would be nice to see some benchmark in the description.

File: python/paddle/nn/functional/norm.py (hunk 1 above)
Comment (huangjun12) [reply python/paddle/nn/functional/norm.py:415]: Done,  functional is referred to class

File: python/paddle/fluid/tests/unittests/test_lrn_op.py (hunk 2 above)
Comment (huangjun12) [reply python/paddle/fluid/tests/unittests/test_lrn_op.py:172]: [3, 40, 40] also works

File: python/paddle/fluid/tests/unittests/test_lrn_op.py (hunk 3 above)
Comment (huangjun12) [reply python/paddle/fluid/tests/unittests/test_lrn_op.py:240]: Yeah, Acc is aligned with torch implementation for simplicity,  here we just test data_format

File: python/paddle/fluid/tests/unittests/test_lrn_op.py (hunk 4 above)
Comment (huangjun12) [reply python/paddle/fluid/tests/unittests/test_lrn_op.py:335]: Done

File: python/paddle/nn/functional/norm.py (hunk 5 above)
Comment (huangjun12) [reply python/paddle/nn/functional/norm.py]: Done

File: python/paddle/nn/functional/norm.py (hunk 6 above)
Comment (huangjun12) [reply python/paddle/nn/functional/norm.py]: Done

**Comment by huangjun12** [comment]
> would be nice to see some benchmark in the description.

 benchmark测试:
|  输入tensor  |            shape               | torch耗时  | paddle耗时 |    
| 3-D Tensor  |       (128, 32, 224)       | 0.01972     | 2.0091        |
| 4-D Tensor  |   (128, 32, 224, 224)   | 1.55033    | 2.0408        |
| 5-D Tensor  | (5, 128, 32, 224, 224) | 9.423509 | 2.13808      |

**Comment by willthefrog** [comment]
> > would be nice to see some benchmark in the description.
> 
> benchmark测试:
> | 输入tensor | shape | torch耗时 | paddle耗时 |
> | 3-D Tensor | (128, 32, 224) | 0.01972 | 2.0091 |
> | 4-D Tensor | (128, 32, 224, 224) | 1.55033 | 2.0408 |
> | 5-D Tensor | (5, 128, 32, 224, 224) | 9.423509 | 2.13808 |

so the time cost of the 3d version is about the same as the 1d version, even though the data is 1120x larger?

**Comment by huangjun12** [comment]
> > > would be nice to see some benchmark in the description.
> > 
> > 
> > benchmark测试:
> > | 输入tensor | shape | torch耗时 | paddle耗时 |
> > | 3-D Tensor | (128, 32, 224) | 0.01972 | 2.0091 |
> > | 4-D Tensor | (128, 32, 224, 224) | 1.55033 | 2.0408 |
> > | 5-D Tensor | (5, 128, 32, 224, 224) | 9.423509 | 2.13808 |
> 
> so the time cost of the 3d version is about the same as the 1d version, even though the data is 1120x larger?

Yes... a little confused

**Comment by willthefrog** [comment]
this is weird, looks like it is implemented with mostly elementwise and reshape operations , would expect it to be linear in complexity.
maybe try increase the input size to find the boundary of the constant time cost?

**Comment by huangjun12** [comment]
> this is weird, looks like it is implemented with mostly elementwise and reshape operations , would expect it to be linear in complexity.
> maybe try increase the input size to find the boundary of the constant time cost?

 the mean cost time of this api is obtained by removing the time of initialization and data copy, results as follows:

| input tensor | shape                          | torch cost  | paddle cost |
| 3-D Tensor  | (32, 224, 224)             | 0.01972     | 0.001512
| 4-D Tensor  | (10, 32, 224, 224)       | 1.55033     | 0.002727
| 5-D Tensor  | (50, 10, 32, 224, 224) | 9.423509  | 0.007594

which is faster than pytorch, and shows  linear complexity as expected.

**Review by jzhang533 — DISMISSED** [review]
lgtm for api change

**Review by willthefrog — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by jzhang533 — APPROVED** [review]
lgtm

**Pull request merged**